### PR TITLE
Fix some of the issues around pgo data collection of stack trace data

### PR DESCRIPTION
**Pull request by davidwrighton**
- Improve assembly name acquisition in dotnet-pgo
  - Use assembly name from AssemblyLoad event, and do not attempt to figure it out from the Module load event
- Collect MemoryRegionInfo for all method regions, not just the first.
- Add a concept to check a guid associated with the associated dll files to avoid data tearing
- Improve parsing of instantiated method signatures when reading an R2R file
  - Handle the subtly different format which is use for type signatures in the InstantiationMethodSignature portion of an R2R file

### CODE REVIEW

File: src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs [hunk 1]
```diff
@@ -47,13 +71,16 @@ internal class MethodMemoryMap
 
                 if (method != null)
                 {
-                    infos.Add(new MemoryRegionInfo
+                    JittedID jittedID = new JittedID(e.MethodID, 0);
+                    if (!info.ContainsKey(jittedID))
```
Comment (jakobbotsch) [inline src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs:75]: Do we see duplicate keys in practice here? Do we know why?

File: src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs [hunk 2]
```diff
@@ -67,25 +94,28 @@ internal class MethodMemoryMap
...
                 }
                 catch
                 {
                 }
 
                 if (method != null)
                 {
-                    infos.Add(new MemoryRegionInfo
+                    JittedID jittedID = new JittedID(e.MethodID, e.ReJITID);
+                    if (!info.ContainsKey(jittedID))
```
Comment (jakobbotsch) [inline src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs:106]: Same question here.

File: src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs [hunk 3]
```diff
@@ -67,25 +94,28 @@ internal class MethodMemoryMap
                 MethodDesc method = null;
                 try
                 {
-                    method = idParser.ResolveMethodID(e.MethodID);
+                    method = idParser.ResolveMethodID(e.MethodID, throwIfNotFound: false);
                 }
                 catch
                 {
                 }
```
Comment (jakobbotsch) [inline src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs:101]: Is the try-catch still necessary?

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs [hunk 4]
```diff
@@ -20,15 +21,38 @@ internal class MethodMemoryMap
         private readonly ulong[] _infoKeys;
         public readonly MemoryRegionInfo[] _infos;
 
+        struct JittedID : IEquatable<JittedID>
```
Comment (EgorBo) [inline src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs:24]: I guess the whole thing can be replace with [record struct](https://sharplab.io/#v2:EYLgtghglgdgPgWAFACcCmBjA9igJgAgGcAXFAVw2PwCkpji1cBJAEQAoAbLGAc3wFk0xABZZmLADT4uvfACU01JgBVWASgDcQA=) if we allowed to use C# 10 in dotnet-pgo 🙂

File: src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs (hunk 1 above)
Comment (davidwrighton) [reply src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs:75]: In theory if the trace contains both MethodLoad and MethodLoadVerbose events, there can be a mess here. In practice though, I haven't seen it.

File: src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs (hunk 3 above)
Comment (davidwrighton) [reply src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs:101]: Sadly, I'm not confident that it never throws. I'm just confident it doesn't throw much with that flag set.

File: src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs (hunk 4 above)
Comment (davidwrighton) [reply src/coreclr/tools/dotnet-pgo/MethodMemoryMap.cs:24]: Probably. I'm still stuck in the C# language I learned years ago, and I forget about all the new stuff.

**Pull request merged**